app/services/order_syncer.rb

```diff
@@ -31,8 +31,19 @@ def sync!
 
   def update_associations_for(order)
     update_bill_address_for(order) if (bill_address.changes.keys & relevant_address_attrs).any?
-    update_ship_address_for(order) if (ship_address.changes.keys & relevant_address_attrs).any?
     update_shipment_for(order) if shipping_method_id_changed?
+
+    # The conditions here are to achieve the same behaviour in earlier versions of Spree, where
+    # switching from pick-up to delivery affects whether simultaneous changes to shipping address
+    # are ignored or not.
+    pickup_to_delivery = force_ship_address_required?(order)
+    if (ship_address.changes.keys & relevant_address_attrs).any?
+      update_ship_address_for(order) unless pickup_to_delivery && order.shipment.blank?
+    end
+    unless pickup_to_delivery && order.shipment.present?
+      order.updater.__send__(:shipping_address_from_distributor)
+    end
+
```

Comment:
this is a bit brain damaging, isnt it? :-)

I didnt manage to improve it much but I think this looks a bit easier to follow:
```ruby
    pickup_to_delivery = force_ship_address_required?(order)
    if !pickup_to_delivery || order.shipment.present?
        update_ship_address_for(order) if (ship_address.changes.keys & relevant_address_attrs).any?
    end
    if !pickup_to_delivery || order.shipment.blank?
      order.updater.__send__(:shipping_address_from_distributor)
    end
```

Reply:
I'd extract this block above including the comment to a separate method, maybe into update_ship_address_for.

Reply:
I see that making this method public would mean touching another class but I think it's worth it.

Reply:
> this is a bit brain damaging, isnt it? :-)

Yes, I had a difficult time following code paths for this. :sweat: 

Thanks @luisramos0 and @mkllnk. I agree with your suggestions and have appended the commits. :slightly_smiling_face: